modules/core/src/smithy4s/http/internals/SchemaVisitorMetadataWriter.scala

```diff
@@ -0,0 +1,207 @@
+            )
+          case _ => MetaEncode.empty
+        }
+      case Primitive.PDocument => MetaEncode.empty[P]
+      case Primitive.PUnit     => MetaEncode.empty[P]
+      case _                   => fromToString[P]
+    }
+  }
+
+  override def list[A](
+      shapeId: ShapeId,
+      hints: Hints,
+      member: Schema[A]
+  ): MetaEncode[List[A]] = {
+    self(member) match {
+      case StringValueMetaEncode(f) =>
+        StringListMetaEncode[List[A]](listA => listA.map(f))
+      case _ => MetaEncode.empty
+    }
+  }
+
+  override def set[A](
+      shapeId: ShapeId,
+      hints: Hints,
+      member: Schema[A]
+  ): MetaEncode[Set[A]] = {
+    self(member) match {
+      case StringValueMetaEncode(f) =>
+        StringListMetaEncode[Set[A]](set => set.map(f).toList)
+      case _ => MetaEncode.empty
+    }
+  }
+
+  override def map[K, V](
+      shapeId: ShapeId,
+      hints: Hints,
+      key: Schema[K],
+      value: Schema[V]
+  ): MetaEncode[Map[K, V]] = {
+    (self(key), self(value)) match {
+      case (StringValueMetaEncode(keyF), StringValueMetaEncode(valueF)) =>
+        StringMapMetaEncode(map =>
+          map.map { case (k, v) =>
+            (keyF(k), valueF(v))
+          }
+        )
+      case (StringValueMetaEncode(keyF), StringListMetaEncode(valueF)) =>
+        StringListMapMetaEncode(map =>
+          map.map { case (k, v) =>
+            (keyF(k), valueF(v))
+          }
+        )
+      case _ => MetaEncode.empty
+    }
+  }
+
+  override def enumeration[E](
+      shapeId: ShapeId,
+      hints: Hints,
+      values: List[EnumValue[E]],
+      total: E => EnumValue[E]
+  ): MetaEncode[E] = {
+    StringValueMetaEncode(e => total(e).stringValue)
+  }
+
+  override def struct[S](
+      shapeId: ShapeId,
+      hints: Hints,
+      fields: Vector[SchemaField[S, _]],
+      make: IndexedSeq[Any] => S
+  ): MetaEncode[S] = {
+    val maybeInputOutput = hints.get[InputOutput]
+    def encodeField[A](
+        field: SchemaField[S, A]
+    ): Option[(Metadata, S) => Metadata] = {
+      val hints = field.hints
+      HttpBinding
+        .fromHints(field.label, hints, maybeInputOutput)
+        .map { binding =>
+          val folderT = new Field.LeftFolder[Schema, Metadata] {
+            override def compile[T](
+                label: String,
+                instance: Schema[T]
+            ): (Metadata, T) => Metadata = {
+              val encoder: MetaEncode[T] =
+                self(
+                  instance.addHints(Hints(binding) ++ hints).addHints(binding)
+                )
+              val updateFunction = encoder.updateMetadata(binding)
+              (metadata, t: T) => updateFunction(metadata, t)
+            }
+          }
+          field.leftFolder(folderT)
+        }
+    }
+    val updateFunctions = fields.map(field => encodeField(field)).collect {
+      case Some(updateFunction) =>
+        updateFunction
+    }
```

Comment:
```suggestion
    val updateFunctions = fields.flatMap(encodeField)
```
Would this be equivalent?

Reply:
I can't remember why I did it this way, might be for 2.12 compatibility but I'm not 100% sure. If Jeff's suggestion works on 2.12 feel free to apply it